File: leetcode/1402-reducing-dishes/reducing-dishes.ts
// 5, 0, -1, -8, -9

function maxSatisfaction(satisfaction: number[]): number {
  satisfaction.sort((a, b) => b - a);

  let result = 0;

  let total = 0;
  for (let i = 0; i < satisfaction.length; i++) {
    if (satisfaction[i] > -total) {
      total += satisfaction[i];
      result += total;
    } else {
      break;
    }
  }

  return result;
}

File: leetcode/1402-reducing-dishes/reducing-dishes.cpp
class Solution {
public:
    int maxSatisfaction(vector<int>& satisfaction) {
        int total = 0;
        int result = 0;

        std::sort(satisfaction.begin(), satisfaction.end(), std::greater<int>());

        for (auto & dish : satisfaction) {
            if (dish > -total) {
                total += dish;
                result += total;
            }
        }

        return result;
    }
};
